File: weathersdk/src/main/java/fr/ekito/myweatherlibrary/di/Module.java
package fr.ekito.myweatherlibrary.di;

import android.app.Application;

/**
 * Inject Module
 * Help define instances
 */
public abstract class Module {

    /**
     * load module definition
     */
    public abstract void load();


    public Application getApplication(){
        return Inject.getApplicationContext();
    }

    /**
     * modules to load
     * @param modules
     */
    public void extend(Class<? extends Module>... modules) {
        for (Class<? extends Module> m : modules) {
            Inject.load(m);
        }
    }

    /**
     * provides several object instances
     * @param instances
     */
    public void provides(Object... instances) {
        for (Object o : instances) {
            Inject.add(o);
        }
    }

    /**
     * provide one object instance
     * @param instance
     * @param clazz
     */
    public void provide(Object instance, Class clazz){
        Inject.add(instance,clazz);
    }
}

File: weathersdk/src/main/kotlin/fr/ekito/myweatherlibrary/di/Module.kt
package fr.ekito.myweatherlibrary.di

import android.app.Application

/**
 * Inject Module
 * Help define instances
 */
abstract class Module {

    /**
     * load module definition
     */
    abstract fun load()

    fun application(): Application? = Inject.applicationContext()

    /**
     * modules to load
     * @param modules
     */
    fun extend(vararg modules: Class<out Module>) {
        for (m in modules) {
            Inject.load(m)
        }
    }

    /**
     * provide one object instance
     * @param instances
     * *
     * @param clazz
     */
    fun provides(vararg instances: Any) {
        instances.forEach { i -> provide(i, i.javaClass) }
    }

    /**
     * provide one object instance
     * @param instance
     * *
     * @param clazz
     */
    fun provide(instance: Any, clazz: Class<*>) {
        Inject.add(instance, clazz)
    }
}
